notifier.go

```diff
@@ -52,32 +60,48 @@ func New(cfg Configuration) (*Notifier, error) { //nolint:gocritic // We want to
 		return nil, err
 	}
 
+	const bufChanSize = 16
+
 	return &Notifier{
 		cfg: &cfg,
 
-		sessionChannel:         make(chan *session, 1),
 		sessions:               []*session{},
-		sessionMutex:           sync.Mutex{},
 		sessionPublishInterval: time.Minute,
+
+		sessionCh: make(chan *session, bufChanSize),
+		reportCh:  make(chan *report, bufChanSize),
+
+		shutdownCh: make(chan struct{}),
 	}, nil
 }
 
+// Close shuts down the notifier, flushing any unsent reports and sessions.
+// Any further calls to StartSession and Notify will panic, so it is your
+// responsibility to only call Close at an appropriate time.
+func (n *Notifier) Close() {
+	// OK, so we have that warning above in the documentation about the panics,
+	// but I'd much rather just drop the sessions/reports. I haven't bothered
+	// figuring out how to do this yet in a clean (race-condition-free) manner.
+	n.shutdownCh <- struct{}{}
+}
+
 // Notify reports the given error to Bugsnag.
 func (n *Notifier) Notify(ctx context.Context, err error) {
 	if err == nil {
 		logErr(errors.New("error missing in call to (*bugsnag.Notifier).Notify. no error reported to Bugsnag"))
 		return
 	}
-	report, ctx := n.makeReport(ctx, err)
+	n.loopOnce.Do(func() { go n.loop() })
+	r := n.makeReport(ctx, err)
 	if sanitizer := n.cfg.ErrorReportSanitizer; sanitizer != nil {
-		ctx = sanitizer(context.Background(), report)
+		ctx = sanitizer(context.Background(), r.report)
```

Comment:
Can we now pass in the `ctx` here instead of `context.Background()`, since the `ctx` var will be whatever was passed in to Notify, and is probably going to give the user the experience they expected. Previously this var would have been the 'deepest' `ctx`, which probably would have led to confusion but now it should be fine I think.

Except!! Right now we're passing the deepest ctx down to the HTTP logic, instead of the user provided/altered one after sanitization.

We should be more explicit about which ctx is which here. Maybe not return a `*report` from the makeReport after all?